packages/babel-core/src/transformation/block-hoist-plugin.js

```diff
@@ -41,23 +39,29 @@ const blockHoistPlugin = {
   visitor: {
     Block: {
       exit({ node }) {
-        let hasChange = false;
-        for (let i = 0; i < node.body.length; i++) {
-          const bodyNode = node.body[i];
-          if (bodyNode?._blockHoist != null) {
-            hasChange = true;
-            break;
+        if (!node.body.find(node => node?._blockHoist != null)) return;
+        // TODO: Babel 9 (?) - remove stabilityMap once Node >= 12 is guaranteed
+        // Array sorting is not stable in earlier Node releases
+        // See: https://v8.dev/blog/array-sort for details
+        const nodeVersion =
+          typeof process !== "undefined"
+            ? process.versions.node.split(".")[0]
+            : undefined;
+        const stabilityMap =
+          nodeVersion && nodeVersion < 12
+            ? new Map(node.body.map((n, idx) => [n, idx]))
```

Comment:
```suggestion
            ? new Map(node.body.entries())
```

Reply:
:100:

Reply:
`Array#entries()` actually gives the reverse of what we need for each element.
```js
const items = ['a', 'b', 'c'];
console.log(new Map(items.map((n, idx) => [n, idx])));
// Map { 'a' => 0, 'c' => 1, 'b' => 2 }
console.log(new Map(items.entries()))
// Map { 0 => 'a', 1 => 'c', 2 => 'b' }
```

We need a Map keyed on the values of `node.body`, not a Map keyed on the indexes.